File: spring-boot/src/main/java/com/inttegro/demo/CheckoutService.java
package com.inttegro.demo;

import com.inttegro.ApiException;
import com.inttegro.Client;
import com.inttegro.RequestMeta;
import com.inttegro.customers.CustomerData;
import com.inttegro.orders.CheckoutSettings;
import com.inttegro.orders.Order;
import com.inttegro.orders.OrderCreateParams;
import com.inttegro.orders.OrderLineItemParams;
import com.inttegro.prices.PriceParams;
import com.inttegro.products.Product;
import com.inttegro.products.ProductType;
import java.net.URI;
import java.util.Locale;

import org.springframework.stereotype.Service;

@Service
public class CheckoutService {
    /*
     * Inttegro integration map
     *
     * INTTEGRO:FLOW [hosted-checkout] This trusted service creates and finalizes
     * the Ledgerline Order and returns the hosted invoice URL.
     * INTTEGRO:SECURITY [server-api-key] INTTEGRO_API_KEY stays in the server
     * process. Never expose it through templates, static assets, client
     * configuration, logs, or exception responses.
     * INTTEGRO:ALTERNATIVE [hosted-checkout] Direct API payment offers more UI
     * control but shifts more payment states, recovery, method-specific behavior,
     * testing, and compliance analysis to the merchant.
     * INTTEGRO:DOCS https://studio.inttegro.com/accept-payment-with-inttegro-checkout
     * INTTEGRO:DOCS https://studio.inttegro.com/orders
     * INTTEGRO:DOCS https://studio.inttegro.com/keys
     * See the repository INTEGRATION_GUIDE.md and integration-decisions.json
     * for the shared rationale and machine-readable alternatives.
     */
    public record CheckoutResult(String orderId, String checkoutUrl) {}
    public record CatalogSelection(ProductType type, String name, String about, String reference, PriceParams price) {}

    public static final class DemoException extends RuntimeException {
        private final String code;

        DemoException(String code, String message) {
            super(message);
            this.code = code;
        }

        public String code() { return code; }
    }

    static CatalogSelection selectCatalogProduct(Product product, String priceId) {
        // INTTEGRO:SECURITY [catalog-authority] The browser supplies no product,
        // price, or amount. Resolve them with the server credential and verify
        // the explicit Price belongs to the configured active Product.
        if (product == null || !Boolean.TRUE.equals(product.active)) {
            throw new DemoException("configuration_error", "The configured demo product is not active.");
        }
        var price = product.prices == null ? null : product.prices.stream()
                .filter(candidate -> priceId.equals(candidate.id))
                .findFirst()
                .orElse(null);
        // This SDK version does not expose the price's active flag in a Product
        // summary. Require an explicit ID match and positive nominal; newer SDKs
        // should also require the returned Price to be active.
        if (price == null || price.nominal == null || price.nominal.value == null || price.nominal.value <= 0) {
            throw new DemoException("configuration_error", "The configured demo price is not available for this product.");
        }
        return new CatalogSelection(
                product.type,
                product.name,
                product.about,
                product.reference,
                PriceParams.of(price.nominal.currency, price.nominal.value));
    }

    static OrderCreateParams buildOrderRequest(CheckoutForm form, String origin, CatalogSelection product) {
        return OrderCreateParams.builder()
                // INTTEGRO:DECISION [stable-idempotency-key] Reuse this key for
                // retries of one logical invoice payment. Production systems
                // should persist an invoice-derived key; a new key after a
                // timeout permits duplicates.
                // https://studio.inttegro.com/idempotency
                .requestMeta(RequestMeta.withIdempotencyKey("demo-" + form.attemptId()))
                // INTTEGRO:DECISION [merchant-order-number] Supply Ledgerline's
                // reference instead of using Inttegro's generated or_ ID as the
                // order number. The validated attempt keeps this demo reference
                // stable across retries. Production should use its persisted
                // invoice number and must not encode customer PII.
                .number("INV-2048-" + form.attemptId().replace('_', '-').toUpperCase(Locale.ROOT).substring(0, Math.min(48, form.attemptId().length())))
                // INTTEGRO:DECISION [inline-customer] customerData fits this
                // guest flow. Account-based apps should resolve customerId on
                // the server; Inttegro accepts exactly one representation.
                .customerData(CustomerData.builder().name(form.name().trim()).email(form.email().trim()).phoneNumber(form.phone().trim()).build())
                // INTTEGRO:DECISION [finalize-on-create] This invoice is settled,
                // so finalization freezes the Order and creates invoice formats.
                // Use draft -> update -> finalize when lines, tax, or approval
                // can still change.
                .finalizeOrder(true)
                .checkoutSettings(CheckoutSettings.builder().redirectUrl(origin + "/complete").cancelUrl(origin + "/cancel").build())
                .lineItem(OrderLineItemParams.product(item -> item
                        // INTTEGRO:DECISION [catalog-snapshot] Resolve the Product
                        // and Price, then snapshot the verified fields into the
                        // Order for compatibility across SDK versions.
                        // INTTEGRO:ALTERNATIVE [catalog-snapshot] When the SDK
                        // exposes the typed catalogue union, send product_id +
                        // price_id + quantity and omit the inline fields.
                        .type(product.type())
                        .name(product.name())
                        .about(product.about())
                        .reference(product.reference())
                        .quantity(1)
                        .price(product.price())))
                .build();
    }

    CheckoutResult create(CheckoutForm form, String origin) {
        // INTTEGRO:SECURITY [server-api-key] Use the deployment secret manager
        // in production and rotate the key there; never serialize it to clients.
        String apiKey = System.getenv().getOrDefault("INTTEGRO_API_KEY", "").trim();
        if (apiKey.isEmpty()) {
            throw new DemoException("configuration_error", "Set INTTEGRO_API_KEY on the server.");
        }
        String productId = System.getenv().getOrDefault("INTTEGRO_DEMO_PRODUCT_ID", "").trim();
        String priceId = System.getenv().getOrDefault("INTTEGRO_DEMO_PRICE_ID", "").trim();
        if (!productId.matches("^prod_[A-Za-z0-9]+$") || !priceId.matches("^pr_[A-Za-z0-9]+$")) {
            throw new DemoException("configuration_error", "Set INTTEGRO_DEMO_PRODUCT_ID and INTTEGRO_DEMO_PRICE_ID on the server.");
        }
        String publicOrigin = validatedOrigin(origin);
        try {
            // INTTEGRO:ALTERNATIVE [server-api-key] A production Spring app can
            // inject a singleton Client for transport reuse, fail-fast config,
            // and application-owned OpenTelemetry. Per-call creation is concise
            // here. https://studio.inttegro.com/sdk-observability
            Client client = new Client(apiKey);
            // INTTEGRO:FLOW [catalog-lookup] Resolve at checkout time so product
            // publication and price changes take effect. High-volume services
            // can use a short cache with explicit invalidation.
            // https://studio.inttegro.com/products
            CatalogSelection product = selectCatalogProduct(client.products().lookup(productId), priceId);
            Order order = client.orders().create(buildOrderRequest(form, publicOrigin, product));
            // INTTEGRO:DECISION [returned-checkout-url] Use the response field.
            // Constructing a URL from order.id relies on undocumented routing.
            if (order.invoice == null || order.invoice.format == null || order.invoice.format.web == null || order.invoice.format.web.url == null) {
                throw new DemoException("api_error", "Inttegro did not return a hosted checkout URL.");
            }
            return new CheckoutResult(order.id, order.invoice.format.web.url);
        } catch (DemoException error) {
            throw error;
        } catch (ApiException error) {
            // INTTEGRO:SECURITY [safe-error-boundary] Keep raw responses,
            // credentials, and stack traces in protected diagnostics. Expose a
            // stable message and record only bounded metadata/request IDs.
            throw new DemoException("api_error", "Inttegro rejected the checkout request.");
        } catch (InterruptedException error) {
            Thread.currentThread().interrupt();
            throw new DemoException("api_error", "Checkout was interrupted. Try again.");
        } catch (Exception error) {
            throw new DemoException("api_error", "Checkout is temporarily unavailable.");
        }
    }

    private static String validatedOrigin(String value) {
        // INTTEGRO:SECURITY [configured-public-origin] Prefer an allow-listed
        // origin in production. A request-derived fallback is safe only behind
        // a precise trusted-proxy configuration. Return URLs must be HTTP(S),
        // and Inttegro appends order_id after checkout.
        try {
            URI uri = URI.create(value);
            if (!("http".equals(uri.getScheme()) || "https".equals(uri.getScheme())) || uri.getHost() == null) throw new IllegalArgumentException();
            return uri.getScheme() + "://" + uri.getAuthority();
        } catch (IllegalArgumentException error) {
            throw new DemoException("configuration_error", "The demo public URL is invalid.");
        }
    }
}

File: laravel/app/Services/CheckoutService.php
<?php

namespace App\Services;

use Inttegro\APIError;
use Inttegro\Client;
use Inttegro\PriceParams;
use Inttegro\Product;
use Inttegro\ProductType;
use RuntimeException;

/*
 * Inttegro integration map
 *
 * INTTEGRO:FLOW [hosted-checkout] This server-only service creates and
 * finalizes the Kora Market Order and returns its hosted invoice URL.
 * INTTEGRO:SECURITY [server-api-key] INTTEGRO_API_KEY belongs only in the
 * trusted PHP process. Never expose it through VITE_ variables, Blade, public
 * assets, logs, or error responses.
 * INTTEGRO:ALTERNATIVE [hosted-checkout] Direct API payment offers more UI
 * control but makes the merchant own more payment state, recovery,
 * method-specific behavior, testing, and compliance analysis.
 * INTTEGRO:DOCS https://studio.inttegro.com/accept-payment-with-inttegro-checkout
 * INTTEGRO:DOCS https://studio.inttegro.com/orders
 * INTTEGRO:DOCS https://studio.inttegro.com/keys
 * See ../../../INTEGRATION_GUIDE.md and ../../../integration-decisions.json for the shared
 * rationale and machine-readable alternatives.
 */

final class DemoError extends RuntimeException
{
    public function __construct(public readonly string $errorCode, string $message, ?\Throwable $previous = null)
    {
        parent::__construct($message, 0, $previous);
    }
}

final class CheckoutService
{
    /**
     * @return array{type: ProductType, name: string, about: ?string, reference: ?string, price: PriceParams}
     */
    public static function catalogSelection(Product $product, string $priceId): array
    {
        // INTTEGRO:SECURITY [catalog-authority] Product, Price, and amount are
        // deployment configuration resolved with the server credential. The
        // public form cannot choose any of them.
        if (!$product->active) {
            throw new DemoError('configuration_error', 'The configured demo product is not active.');
        }

        $selected = null;
        foreach ($product->prices ?? [] as $price) {
            if ($price->id === $priceId && $price->active) {
                $selected = $price;
                break;
            }
        }
        if ($selected === null || $selected->nominal->value <= 0) {
            throw new DemoError('configuration_error', 'The configured demo price is not active for this product.');
        }

        return [
            'type' => ProductType::from($product->type),
            'name' => $product->name,
            'about' => $product->about,
            'reference' => $product->reference,
            'price' => new PriceParams($selected->nominal->currency, $selected->nominal->value),
        ];
    }

    /**
     * @param array{name: string, email: string, phone: string, attempt_id: string} $checkout
     * @param array{type: ProductType, name: string, about: ?string, reference: ?string, price: PriceParams} $product
     */
    public static function orderPayload(array $checkout, string $origin, array $product): array
    {
        return [
            // INTTEGRO:DECISION [stable-idempotency-key] Reuse the key for a
            // retry of this logical checkout. Production code should persist a
            // cart/invoice-derived key; creating a new key after a timeout can
            // create a duplicate. https://studio.inttegro.com/idempotency
            'request_meta' => ['idempotency_key' => 'demo-'.$checkout['attempt_id']],
            // INTTEGRO:DECISION [merchant-order-number] Supply a recognizable
            // merchant reference instead of accepting Inttegro's generated or_
            // ID fallback. The validated attempt makes this demo value stable
            // across retries. Production should use its persisted sales-order
            // number and never include customer PII.
            'number' => 'KORA-'.strtoupper(substr(str_replace('_', '-', $checkout['attempt_id']), 0, 48)),
            // INTTEGRO:DECISION [inline-customer] customer_data suits this guest
            // flow. Account-based apps should resolve customer_id server-side;
            // the Orders API accepts exactly one customer representation.
            'customer_data' => ['name' => $checkout['name'], 'email_address' => $checkout['email'], 'phone_number' => $checkout['phone']],
            // INTTEGRO:DECISION [finalize-on-create] The cart is settled, so one
            // operation freezes the Order and produces invoice formats. Use a
            // draft, updates, then finalization when tax, shipping, inventory,
            // approval, or line items can still change.
            'finalize' => true,
            'checkout_settings' => ['redirect_url' => $origin.'/complete', 'cancel_url' => $origin.'/cancel'],
            'line_items' => [[
                'type' => 'product',
                'product' => [
                    // INTTEGRO:DECISION [catalog-snapshot] The service looks up
                    // the configured Product and Price, verifies their
                    // relationship, and snapshots those authoritative fields
                    // into the Order for compatibility across SDK versions.
                    // INTTEGRO:ALTERNATIVE [catalog-snapshot] When the SDK
                    // exposes the typed catalogue union, send product_id +
                    // price_id + quantity instead and omit the inline fields.
                    'type' => $product['type'],
                    'name' => $product['name'],
                    'about' => $product['about'],
                    'reference' => $product['reference'],
                    'quantity' => 1,
                    'price' => $product['price'],
                ],
            ]],
        ];
    }

    /** @param array{name: string, email: string, phone: string, attempt_id: string} $checkout
     *  @return array{order_id: string, checkout_url: string}
     */
    public static function create(array $checkout, string $origin): array
    {
        // INTTEGRO:SECURITY [server-api-key] Load the key from a deployment
        // secret store in production and rotate it there. Do not serialize it.
        $apiKey = trim((string) getenv('INTTEGRO_API_KEY'));
        if ($apiKey === '') {
            throw new DemoError('configuration_error', 'Set INTTEGRO_API_KEY on the server.');
        }
        $productId = trim((string) getenv('INTTEGRO_DEMO_PRODUCT_ID'));
        $priceId = trim((string) getenv('INTTEGRO_DEMO_PRICE_ID'));
        if (!preg_match('/^prod_[A-Za-z0-9]+$/', $productId) || !preg_match('/^pr_[A-Za-z0-9]+$/', $priceId)) {
            throw new DemoError('configuration_error', 'Set INTTEGRO_DEMO_PRODUCT_ID and INTTEGRO_DEMO_PRICE_ID on the server.');
        }
        // INTTEGRO:SECURITY [configured-public-origin] Prefer an allow-listed
        // origin in production. Request-derived values are safe only behind a
        // precise trusted-proxy configuration. Return URLs must be HTTP(S), and
        // Inttegro appends order_id after checkout.
        $parts = parse_url($origin);
        if (!is_array($parts) || !in_array($parts['scheme'] ?? '', ['http', 'https'], true) || empty($parts['host'])) {
            throw new DemoError('configuration_error', 'The demo public URL is invalid.');
        }
        $publicOrigin = $parts['scheme'].'://'.$parts['host'].(isset($parts['port']) ? ':'.$parts['port'] : '');

        try {
            // INTTEGRO:ALTERNATIVE [server-api-key] A production container may
            // inject one startup-configured Client for transport reuse and
            // application-owned OpenTelemetry. Per-call construction is concise
            // for the demo. https://studio.inttegro.com/sdk-observability
            $client = new Client($apiKey);
            // INTTEGRO:FLOW [catalog-lookup] Resolve the Product at checkout so
            // publication and price changes take effect. High-volume services
            // can use a short cache with explicit invalidation.
            // https://studio.inttegro.com/products
            $product = self::catalogSelection($client->products->lookup($productId), $priceId);
            $order = $client->orders->create(self::orderPayload($checkout, $publicOrigin, $product));
            // INTTEGRO:DECISION [returned-checkout-url] Use the URL in the
            // response. Constructing one from the order ID relies on
            // undocumented routing conventions.
            $checkoutUrl = $order->invoice?->format?->web?->url;
            if (!$checkoutUrl) {
                throw new DemoError('api_error', 'Inttegro did not return a hosted checkout URL.');
            }
            return ['order_id' => $order->id, 'checkout_url' => $checkoutUrl];
        } catch (DemoError $error) {
            throw $error;
        } catch (APIError $error) {
            // INTTEGRO:SECURITY [safe-error-boundary] Keep raw API bodies,
            // credentials, and traces in protected diagnostics. Public messages
            // remain stable; log only bounded metadata and request IDs.
            throw new DemoError('api_error', 'Inttegro rejected the checkout request.', previous: $error);
        } catch (\Throwable $error) {
            throw new DemoError('api_error', 'Checkout is temporarily unavailable.', previous: $error);
        }
    }
}
